querysrv.go

```diff
@@ -3,213 +3,305 @@
 	s.limiter = lru.New(lruSize)
 
-	conn, err := net.ListenUDP("udp", s.addr)
+	tlsCfg := &tls.Config{
+		Certificates:           []tls.Certificate{s.cert},
+		ClientAuth:             tls.RequestClientCert,
+		SessionTicketsDisabled: true,
+		MinVersion:             tls.VersionTLS12,
+		CipherSuites: []uint16{
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+		},
+	}
+
+	http.HandleFunc("/", s.handler)
+
+	tlsListener, err := tls.Listen("tcp", s.addr, tlsCfg)
 	if err != nil {
 		log.Println("Listen:", err)
 		return
 	}
 
-	// Attempt to set the read and write buffers to 2^24 bytes (16 MB) or as high as
-	// possible.
-	for i := 24; i >= 16; i-- {
-		if conn.SetReadBuffer(1<<uint(i)) == nil {
-			break
-		}
+	s.listener = tlsListener
+
+	srv := &http.Server{
+		ReadTimeout:    30 * time.Second,
+		WriteTimeout:   10 * time.Second,
+		MaxHeaderBytes: 2 << 10,
 	}
-	for i := 24; i >= 16; i-- {
-		if conn.SetWriteBuffer(1<<uint(i)) == nil {
-			break
-		}
+
+	if err := srv.Serve(tlsListener); err != nil {
+		log.Println("Serve:", err)
 	}
+}
 
-	var buf = make([]byte, 1024)
-	for {
-		buf = buf[:cap(buf)]
-		n, addr, err := conn.ReadFromUDP(buf)
-		if err != nil {
-			log.Println("Read:", err)
-			return
-		}
+func (s *querysrv) handler(w http.ResponseWriter, req *http.Request) {
+	if debug {
+		log.Println(req.Method, req.URL)
+	}
 
-		if s.limit(addr) {
-			// Rate limit in effect for source
-			continue
-		}
+	remoteAddr, err := net.ResolveTCPAddr("tcp", req.RemoteAddr)
+	if err != nil {
+		log.Println("remoteAddr:", err)
+		http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+		return
+	}
 
-		if n < 4 {
-			log.Printf("Received short packet (%d bytes)", n)
-			continue
+	if s.limit(remoteAddr.IP) {
+		if debug {
+			log.Println(remoteAddr.IP, "is limited")
 		}
+		w.Header().Set("Retry-After", "60")
+		http.Error(w, "Too Many Requests", 429)
+		return
+	}
 
-		buf = buf[:n]
-		magic := binary.BigEndian.Uint32(buf)
-
-		switch magic {
-		case discover.AnnouncementMagic:
-			err := s.handleAnnounce(addr, buf)
-			globalStats.Announce()
-			if err != nil {
-				log.Println("Announce:", err)
-				globalStats.Error()
-			}
-
-		case discover.QueryMagic:
-			err := s.handleQuery(conn, addr, buf)
-			globalStats.Query()
-			if err != nil {
-				log.Println("Query:", err)
-				globalStats.Error()
-			}
-
-		default:
-			globalStats.Error()
-		}
+	switch req.Method {
+	case "GET":
+		s.handleGET(w, req)
+	case "POST":
+		s.handlePOST(w, req)
+	default:
+		globalStats.Error()
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
 	}
 }
 
-func (s *querysrv) Stop() {
-	panic("stop unimplemented")
-}
+func (s *querysrv) handleGET(w http.ResponseWriter, req *http.Request) {
+	if req.TLS == nil {
```

Comment:
This talks about wether or not the client provided a cert. I'd like to be able to easily ask people to use curl to debug stuff, without asking them to generate a key pair.